doc/design/memory.md

```diff
@@ -0,0 +1,1595 @@
+the associated space metadata, given a value pointer.  A global radix tree
+maintains adequate state to perform this operation.  In order to limit radix
+tree overhead, spaces conform to special size and alignment constraints.  The
+space allocation quantum is a chosen power of 2, e.g. 2 MiB (2<sup>21</sup>).
+All spaces are sized as a whole multiple of the quantum, and they are aligned to
+a quantum boundary.  This allows the radix tree to ignore the least significant
+lg(quantum) address bits, and each space typically only requires one entry in
+the radix tree.  (The extremely unusual exception is for a multi-quantum space
+that contains a huge value, and interior references to that value allow
+references to span multiple quanta.)
+
+### Generation coalescing
+
+The nursery generation grows until its space cannot serve an allocation request,
+at which point it becomes an aging generation.  Aside from allocation via lazy
+suspension execution, a generation cannot grow after it transitions to aging.
+Meanwhile new generations continue to stack up as nurseries transition to aging.
+As generations age, they tend to shrink, but in many cases they will linger due
+to long-lived allocations.  We coalesce adjacent aging generations once they
+shrink below sizes that allow two generations to fit in the space that one
+generation originally required, and their combined utilization drops below the
+copy-collection threshold.  This assures that the generation count remains a
+function of global heap size, rather than growing indefinitely as a function of
+global allocation volume.
+
+### Incremental work
+
+Global collection is a continuous process enabled by various incremental work,
+some explicit, and some that requires intentional asynchronous scheduling.  As
+an example of explicit work, recall that value allocation (a side effect of
+message send) requires the payload to be transitively advanced.  Asynchronous
+incremental work can be hooked to various events, especially:
+
+- Nursery creation
+- Actor switch
+- Value allocation (message send)
+- Value deallocation
+
+There are several types of incremental work:
+
+- Process an element in an `unreachable` value stack.
+- Copy a value fromspace->tospace.
+- Advance an actor.
+- Advance a generation's space(s).
+
+There is no obvious fundamental principle on which to base pace of advancement,
+except perhaps nursery creation rate and/or impending memory exhaustion.  On the
+other hand, decrefing and copying can be driven at a rate proportional to
+allocation rate that is sufficient to guarantee this work does not fall
+hopelessly behind allocation.
+
+Copying is triggered at a fixed utilization threshold, so the mean number of
+times a value can be copied is fixed.  For example, if minimum utilization is
+0.5, the total number of bytes copied per byte allocated is approximately the
+integral of 0.5<sup>x</sup> in the range 0..∞, which is ~1.44.  Therefore the
```

Comment:
Neat, I'd actually been wondering how much we could possibly copy. It's cool that the answer was so simple.

Reply:
It will be interesting to see what utilization range ends up being practical.  I hope we can use something much higher than 0.5, but even 0.5 won't break us.